feat(secrethub) add org_id to openidconnect tokens

colemickens opened the pull request:
## 📝 Description

Add `org_id` to the OIDC Tokens issues in pipeline executions. Fixes: https://github.com/semaphoreio/semaphore/issues/204

If docs are co-located in this repo, I can add them too... 

I'm not sure you want the third commit at all, and if you do I suspect you'll want to tweak wording:
[ ] not sure how you're handling documentation of features that are limited in hosted variants
[ ] not sure what the right branding is "ScaleUp"? "Semaphone {Hosted,Cloud}" to disambiguate from self-hosted installs.
[ ] not actually sure if this feature is easily available in self-hosted, or if it's still semi-blocked-off?

## ✅ Checklist
- [ ] I have tested this change
- [x] This change requires documentation update

colemickens commented:
Keeping this as draft until I can do some local testing...

DamjanBecirovic commented:
Hi @colemickens, sorry for a bit slower response.

The easiest way to run local tests for this feature is to comment out [these two lines](https://github.com/semaphoreio/semaphore/blob/main/secrethub/docker-compose.yml#L21-L22) and then run `make test.ex` in the secrethub directory. This assumes that you have `docker` and `docker-compose` installed locally.

Regarding the questions about docs, the best option would be to use the same approach as [here](https://github.com/semaphoreio/semaphore/blob/main/docs/docs/using-semaphore/openid.md?plain=1#L17) for adding the note about availability on different plans. 

One additional thing, we have a restriction that blocks us from merging commits that are not signed. Can you sign these commits? You can find more information about that [here](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits).

And finally, please also add the key-value pair for `org_id` [here](https://github.com/semaphoreio/semaphore/blob/main/secrethub/lib/secrethub/open_id_connect/jwt_claim.ex#L105) so users can configure its presence in the token from the UI. Other than name and description, you can use the same values as for [wf_id](https://github.com/semaphoreio/semaphore/blob/main/secrethub/lib/secrethub/open_id_connect/jwt_claim.ex#L124-L131) key.

colemickens commented:
Thanks @DamjanBecirovic.

0. I fixed up the doc change, added `org_id` to the UI, and signed the commits and pushed.
1. I'm trying to see about writing a test, but I haven't found any that actually assert the claim values to copy.
2. I'd like to be able to either write a test that dumps/asserts the `org_id` to validate my assumptions, or be able to `docker run` semaphore itself and test a user/project/etc, but I suspect that might be rather involved?
3. Regardless, local testing is not going well anyway:

```
cole@ds-ws-colemickens ~/work/code/semaphore/secrethub @f7fa02bd*
❯ PAGER=cat jj diff --git
diff --git a/secrethub/docker-compose.yml b/secrethub/docker-compose.yml
index 6c32153476..db458228fa 100644
--- a/secrethub/docker-compose.yml
+++ b/secrethub/docker-compose.yml
@@ -18,8 +18,8 @@
     ports:
       - "50051:50051"
     tty: true
-    volumes:
-      - .:/app
+    # volumes:
+    #   - .:/app

     links:
       - db:db
```
```
cole@ds-ws-colemickens ~/work/code/semaphore/secrethub @f7fa02bd*
❯ nix run github:nixos/nixpkgs/nixos-unstable#gnumake -- test.ex
docker compose -f docker-compose.yml build --build-arg BUILDKIT_INLINE_CACHE=1 --build-arg BUILD_ENV=test app
WARN[0000] /home/cole/work/code/semaphore/secrethub/docker-compose.yml: the attribute `version` is obsolete, it will be ignored, please remove it to avoid potential confusion
Compose can now delegate builds to bake for better performance.
 To do so, set COMPOSE_BAKE=true.
[+] Building 11.0s (8/23)                                                                                                     docker:default
 => [app internal] load build definition from Dockerfile                                                                                0.0s
 => => transferring dockerfile: 2.07kB                                                                                                  0.0s
 => [app internal] load metadata for docker.io/hexpm/elixir:1.13.4-erlang-24.3.4.9-alpine-3.18.6                                        0.3s
 => [app internal] load .dockerignore                                                                                                   0.0s
 => => transferring context: 2B                                                                                                         0.0s
 => ERROR [app] importing cache manifest from local/secrethub/main:test                                                                 0.5s
 => [app internal] load build context                                                                                                   0.0s
 => => transferring context: 12.99kB                                                                                                    0.0s
 => [app base  1/14] FROM docker.io/hexpm/elixir:1.13.4-erlang-24.3.4.9-alpine-3.18.6@sha256:8b7fba6311a89a303f176d2f8ba3a2ad69895b23c  0.0s
 => CACHED [app base  2/14] RUN echo "Build for test environment started"                                                               0.0s
 => ERROR [app base  3/14] RUN apk update && apk add --no-cache build-base git python3 curl openssh                                    10.2s
------
 > [app] importing cache manifest from local/secrethub/main:test:
------
------
 > [app base  3/14] RUN apk update && apk add --no-cache build-base git python3 curl openssh:
0.153 fetch https://dl-cdn.alpinelinux.org/alpine/v3.18/main/x86_64/APKINDEX.tar.gz
5.158 WARNING: updating and opening https://dl-cdn.alpinelinux.org/alpine/v3.18/main: temporary error (try again later)
5.158 fetch https://dl-cdn.alpinelinux.org/alpine/v3.18/community/x86_64/APKINDEX.tar.gz
10.16 WARNING: updating and opening https://dl-cdn.alpinelinux.org/alpine/v3.18/community: temporary error (try again later)
10.16 4 unavailable, 0 stale; 26 distinct packages available
------
failed to solve: process "/bin/sh -c apk update && apk add --no-cache build-base git python3 curl openssh" did not complete successfully: exit code: 4
make: *** [../Makefile:216: test.ex] Error 1

```

dexyk commented:
Hi @colemickens! 
Thanks so much for your contribution and for the detailed feedback! It’s really appreciated and helps us improve the dev experience for everyone.

1. For writing a test, you might want to look at the assertions [here](https://github.com/semaphoreio/semaphore/blob/main/secrethub/test/secrethub/internal_grpc_api_test.exs#L833) or [here](https://github.com/semaphoreio/semaphore/blob/main/secrethub/test/secrethub/internal_grpc_api_test.exs#L878). [This section](https://github.com/semaphoreio/semaphore/blob/main/secrethub/test/secrethub/internal_grpc_api_test.exs#L798-L1000) covers token content.

2. Running Semaphore locally is a bit involved right now. For now, focused unit/integration tests are your best bet.

3. Regarding the Docker build error
```
5.158 WARNING: updating and opening https://dl-cdn.alpinelinux.org/alpine/v3.18/main: temporary error (try again later)
```
looks like a temporary network issue reaching Alpine’s package mirror. We couldn’t reproduce it, so it might help to just retry the build.

Let us know if you hit any other issues - happy to help!

colemickens commented:
@dexyk Hi, thanks for the reply. I tried repeatedly over the course of days to get the Docker tests to run, without success - Googling indicates it's some interaction of Docker and Tailscale? Docker + networking, a classic duo.

Unfortunately, I lack both the time and patience to deal with Docker, and I don't know Elixir at all.

Additionally, as I indicated in #204, it would be prudent to include `org` and `prj` claims as well, and I think those are less trivial - from what I understand that data might not already be in `req`, and would likely require changes to more tests to have that data available/asserted upon.

dexyk commented:
Hi @colemickens, thanks again for your effort!

Have you tried setting custom DNS servers in Docker by editing `/etc/docker/daemon.json`?

```json
{
  "dns": ["1.1.1.1", "8.8.8.8"]
}
```

Then restart Docker:

```bash
sudo systemctl restart docker
```

This should help if the issue is related to DNS resolution inside containers (e.g. if Tailscale messed it up). Let us know if this approach works for you, if not, we’ll suggest something else.

colemickens commented:
Same result. Yay docker.

```
cole@ds-ws-colemickens ~/work/code/semaphore/secrethub @f7fa02bd* 38s
❯ cat /etc/docker/daemon.json
{
  "dns": [ "1.1.1.1", "8.8.8.8" ]
}

cole@ds-ws-colemickens ~/work/code/semaphore/secrethub @f7fa02bd*
❯ sudo systemctl restart docker

cole@ds-ws-colemickens ~/work/code/semaphore/secrethub @f7fa02bd*
❯ nix run github:nixos/nixpkgs/nixos-unstable#gnumake -- test.ex
docker compose -f docker-compose.yml build --build-arg BUILDKIT_INLINE_CACHE=1 --build-arg BUILD_ENV=test app
WARN[0000] /home/cole/work/code/semaphore/secrethub/docker-compose.yml: the attribute `version` is obsolete, it will be ignored, please remove it to avoid potential confusion
Compose can now delegate builds to bake for better performance.
 To do so, set COMPOSE_BAKE=true.
[+] Building 11.2s (8/23)                                                docker:default
 => [app internal] load build definition from Dockerfile                           0.0s
 => => transferring dockerfile: 2.07kB                                             0.0s
 => [app internal] load metadata for docker.io/hexpm/elixir:1.13.4-erlang-24.3.4.  0.6s
 => [app internal] load .dockerignore                                              0.0s
 => => transferring context: 2B                                                    0.0s
 => ERROR [app] importing cache manifest from local/secrethub/main:test            0.3s
 => [app base  1/14] FROM docker.io/hexpm/elixir:1.13.4-erlang-24.3.4.9-alpine-3.  0.0s
 => [app internal] load build context                                              0.0s
 => => transferring context: 12.99kB                                               0.0s
 => CACHED [app base  2/14] RUN echo "Build for test environment started"          0.0s
 => ERROR [app base  3/14] RUN apk update && apk add --no-cache build-base git p  10.2s
------
 > [app] importing cache manifest from local/secrethub/main:test:
------
------
 > [app base  3/14] RUN apk update && apk add --no-cache build-base git python3 curl openssh:
0.155 fetch https://dl-cdn.alpinelinux.org/alpine/v3.18/main/x86_64/APKINDEX.tar.gz
5.160 fetch https://dl-cdn.alpinelinux.org/alpine/v3.18/community/x86_64/APKINDEX.tar.gz
5.160 WARNING: updating and opening https://dl-cdn.alpinelinux.org/alpine/v3.18/main: temporary error (try again later)
10.16 4 unavailable, 0 stale; 26 distinct packages available
10.16 WARNING: updating and opening https://dl-cdn.alpinelinux.org/alpine/v3.18/community: temporary error (try again later)
------
failed to solve: process "/bin/sh -c apk update && apk add --no-cache build-base git python3 curl openssh" did not complete successfully: exit code: 4
make: *** [../Makefile:216: test.ex] Error 1

```

colemickens commented:
That DNS doesn't seem to be respected by docker. I don't know if I'm doing something wrong. When I `docker run -it ubuntu` and check `/etc/resolv.conf`, there's clearly a `10.100.0.2` nameserver. No idea where that comes from, or why Docker is injecting that.

dexyk commented:
@colemickens, if you’re up for tweaking your host environment a bit more, another potential fix is disabling Tailscale’s stateful filtering. Folks in this Docker-yt-dlp thread saw success: https://github.com/jim60105/docker-yt-dlp/issues/2, and the resolution comment here: https://github.com/tailscale/tailscale/issues/12108#issuecomment-2106489435.

You can try:

```bash
sudo tailscale up --stateful-filtering=false
```

Then restart Docker and rerun your build.

colemickens commented:
@dexyk As I noted in #204, for the sake of consistency in Semaphore, the product, and to facilitate integrations like the one we are proposing, I think that `org` and `prj` claims should be added as well.

Otherwise I can go ahead and rebase this and mark it ready, sure.

dexyk commented:
Hi @colemickens, we’d like to go ahead and merge the current changes as they are, really appreciate all the effort you put in.

Let’s keep the `org/proj` claims discussion going in [the linked GitHub issue](https://github.com/semaphoreio/semaphore/issues/204) (I added a [reply](https://github.com/semaphoreio/semaphore/issues/204#issuecomment-2841303058) there). Let me know what you think.

dexyk commented:
Hi @colemickens, we have a policy requiring signed commits before merging. Could you please sign yours? Thanks!

AleksandarCole commented:
Merging with approval from @darkofabijan
